### ISSUE TITLE
Multi instance pgsql

### ISSUE THREAD

**Issue by inigosalvat** [ClusterLabs/resource-agents#591]
Hi,

I'm experiencing problems for an environment with two postgresql 9.3 instances in a hot standby mode.

I am able to setup each of them stand alone (each instance has its own port and virtual IP) but when I try to setup both at the same time (the two master in the same node or in different one) only one of them starts properly (I mean that starts both nodes and synchronize them via streaming).

Does any of you have test that? Do you have a simple guide or  any trick I must take care?

Thanks for your help,

Iñigo

**Comment by kskmori** [comment]
At least we've tested once and made a necessary fix for a multi-instance configuration at #118 , which is included as of v3.9.4.
If you're using the older one, please try upgrade it. If you still see a problem with the latest one, providing your crm configuration and the logs would be helpful.

@knakahira : I'd appreciate if you could take a look the logs when provided.

**Comment by dmuhamedagic** [comment]
@inigosalvat Could you provide the logs?

**Comment by ateyk** [comment]
It's look that we faced with similar issue.
Here is our steps:
1. 6 nodes cluster (a1, a3, a5, a6, a7, b1);
2. One pgsql cluster instance on a5 and a7 (id pgsql). This cluster configured as following:

> pcs resource show pgsql
>  Resource: pgsql (class=ocf provider=heartbeat type=pgsql)
>   Attributes: pgctl=/usr/pgsql-9.4/bin/pg_ctl psql=/usr/pgsql-9.4/bin/psql pgdata=/var/lib/pgsql/9.4/data pglibs=/usr/pgsql-9.4/lib rep_mode=async node_list="a5 a7" master_ip=10.0.0.10 restart_on_promote=true replication_slot_name=slave tmpdir=/var/lib/pgsql/tmp
>   Meta Attrs: migration-threshold=5 failure-timeout=120
>   Operations: start on-fail=restart interval=0s timeout=60 (pgsql-start-interval-0s)
>               monitor on-fail=restart interval=20 timeout=60 (pgsql-monitor-interval-20)
>               monitor interval=15 role=Master timeout=60 (pgsql-monitor-interval-15)
>               promote on-fail=restart interval=0s timeout=30 (pgsql-promote-interval-0s)
>               demote on-fail=stop interval=0s timeout=30 (pgsql-demote-interval-0s)
>               stop on-fail=block interval=0s timeout=60 (pgsql-stop-interval-0s)
>               notify interval=0s timeout=60 (pgsql-notify-interval-0s)

 
with additional constraints that bans other nodes except a5 and a7
This resource works quite nice (with failover)
3. Trying setup new pgsql cluster instance on a1 and a6 (id pgsql_n)

> pcs resource show pgsql_n
>  Resource: pgsql_n (class=ocf provider=heartbeat type=pgsql)
>   Attributes: pgctl=/usr/pgsql-9.4/bin/pg_ctl psql=/usr/pgsql-9.4/bin/psql pgdata=/var/lib/pgsql/9.4/data pglibs=/usr/pgsql-9.4/lib rep_mode=async node_list="a1 a6" master_ip=10.0.0.13 restart_on_promote=true replication_slot_name=slave tmpdir=/var/lib/pgsql/tmp
>   Meta Attrs: migration-threshold=5 failure-timeout=120
>   Operations: start on-fail=restart interval=0s timeout=60 (pgsql_n-start-interval-0s)
>               monitor on-fail=restart interval=20 timeout=60 (pgsql_n-monitor-interval-20)
>               monitor interval=15 role=Master timeout=60 (pgsql_n-monitor-interval-15)
>               promote on-fail=restart interval=0s timeout=30 (pgsql_n-promote-interval-0s)
>               demote on-fail=stop interval=0s timeout=30 (pgsql_n-demote-interval-0s)
>               stop on-fail=block interval=0s timeout=60 (pgsql_n-stop-interval-0s)
>               notify interval=0s timeout=60 (pgsql_n-notify-interval-0s)

with additional constraints that bans other nodes except a1 and a6
4. After pcs cib-push pg.cib (with added pgsql_n) things goes bad.
On a5

> pgsql(pgsql_n)[23174]:	2016/07/15_22:28:51 INFO: Changing pgsql_n-status on a5 : ->PRI.
> Jul 15 22:28:51 [1996] a5      attrd:     info: attrd_peer_update:	Setting pgsql_n-status[a5]: (null) -> PRI from a5
> 

On a7

> pgsql(pgsql_n)[3994]:	2016/07/15_22:28:51 INFO: Changing pgsql_n-status on a7 : ->HS:alone.
> Jul 15 22:28:51 [19517] a7      attrd:     info: attrd_peer_update:	Setting pgsql_n-status[a7]: (null) -> HS:alone from a7
> 

After that failover behaviour happen: a5 fails and a7 become new master, a1 and a6 stoped.

I think it's strange, that a5 and a7 tried to do something with pgsql_n.
I will provide more logging if needed (it's not available at the moment)

**Comment by kskmori** [comment]
Hi @ateyk ,

In your case, the issue comes from both instances has the same pgdata=/var/lib/pgsql/9.4/data . 
The pgdata is the key to distinguish each instances so they must be different locations for each instances.

**Comment by ateyk** [comment]
@kskmori, thank you for fast response, it's now clear for me.
I see that my trouble not connected with original issue, sorry.

**Comment by ateyk** [comment]
Oh, would you please clarify one more thing, is you suggestion actual even if instances resides on different nodes and has different ids?

**Comment by kskmori** [comment]
Yes, to the best of my knowledge they needs to be different even if instances resides on different nodes.

It might be possible to use the same value by utilizing 'resource-discovery' feature of Pacemaker, but it's only my guess and I have never verified if it works for the pgsql multi-instance configuration.
